plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -319,6 +319,136 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	}
 	return (guint16)val;
 }
+static gboolean
+fu_thunderbolt_check_usb4_port_path(FuUdevDevice *device, const gchar *attribute, GError **err)
+{
+	g_autofree const gchar *path =
+	    g_build_filename(fu_udev_device_get_sysfs_path(device), attribute, NULL);
+	g_autofree gchar *fn = g_strdup_printf("%s", path);
+	g_autoptr(GFile) file = g_file_new_for_path(fn);
+	if (!g_file_query_exists(file, NULL)) {
+		g_set_error(err, G_IO_ERROR, G_IO_ERROR_NOT_FOUND, "failed to find %s", fn);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device, GError **err)
+{
+	const gchar *offline = "usb4_port1/offline";
+	const gchar *rescan = "usb4_port1/rescan";
+	GError **err_usb4_port_offline_path = NULL;
+	GError **err_usb4_port_rescan_path = NULL;
+	if (fu_thunderbolt_check_usb4_port_path(device, offline, err_usb4_port_offline_path)) {
+		if (!fu_udev_device_write_sysfs(device, offline, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Setting usb4 port offline failed");
+			return FALSE;
+		}
+	} else {
+		return TRUE;
+	}
+	if (fu_thunderbolt_check_usb4_port_path(device, rescan, err_usb4_port_rescan_path)) {
+		if (!fu_udev_device_write_sysfs(device, rescan, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Rescan on port failed");
+			return FALSE;
+		}
+	} else {
+		return TRUE;
```

Comment:
just a nit, you may want `g_debug("%s", err_usb4_port_rescan_path->message);` here.